File: app/src/main/java/com/remotearthsolutions/expensetracker/databaseutils/daos/CategoryDao.java
package com.remotearthsolutions.expensetracker.databaseutils.daos;

import androidx.lifecycle.LiveData;
import androidx.room.*;
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel;
import com.remotearthsolutions.expensetracker.databaseutils.models.dtos.CategoryExpense;
import io.reactivex.Flowable;

import java.util.List;

@Dao
public interface CategoryDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void addCategory(CategoryModel categoryModel);

    @Insert
    void addAllCategories(CategoryModel... categoryModels);

    @Query("Select * from category")
    Flowable<List<CategoryModel>> getAllCategories();

    @Delete
    void deleteCategory(CategoryModel categoryModel);

    @Query("DELETE FROM category")
    void deleteAll();

    @Update
    void updateCategory(CategoryModel categoryModel);

    @Query("SELECT COUNT(category_name) FROM category")
    LiveData<Integer> getDataCount();
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/databaseutils/daos/CategoryDao.kt
package com.remotearthsolutions.expensetracker.databaseutils.daos

import androidx.lifecycle.LiveData
import androidx.room.*
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel
import io.reactivex.Flowable

@Dao
interface CategoryDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun addCategory(categoryModel: CategoryModel)

    @Insert
    fun addAllCategories(vararg categoryModels: CategoryModel)

    @get:Query("Select * from category")
    val allCategories: Flowable<List<CategoryModel>>

    @Delete
    fun deleteCategory(categoryModel: CategoryModel)

    @Query("DELETE FROM category")
    fun deleteAll()

    @Update
    fun updateCategory(categoryModel: CategoryModel?)

    @get:Query("SELECT COUNT(category_name) FROM category")
    val dataCount: LiveData<Int>
}